runtime/src/status_cache.rs

```diff
@@ -1,411 +1,261 @@
+                trace!("get_signature_status_slow: got {}", forkid);
+                return ancestors
+                    .get(&forkid)
+                    .map(|id| (*id, res.clone()))
+                    .or_else(|| Some((ancestors.len(), res)));
             }
         }
         None
     }
-    pub fn get_signature_status(&self, sig: &Signature) -> Option<Result<(), T>> {
-        self.active
-            .as_ref()
-            .and_then(|active| active.get_signature_status(sig))
-            .or_else(|| self.get_signature_status_merged(sig))
-    }
 
-    fn squash_parent_is_full(&mut self, parent: &Self) -> bool {
-        // flatten and squash the parent and its merges into self.merges,
-        //  returns true if self is full
-        if parent.active.is_some() {
-            warn!("=========== FIXME: squash() on an active parent! ================");
-        }
-        // TODO: put this assert back in
-        //assert!(parent.active.is_none());
-
-        if self.merges.len() < MAX_CACHE_ENTRIES {
-            for merge in parent
-                .merges
-                .iter()
-                .take(MAX_CACHE_ENTRIES - self.merges.len())
-            {
-                self.merges.push_back(merge.clone());
-            }
-        }
-        self.merges.len() == MAX_CACHE_ENTRIES
+    pub fn register_hash(&mut self, blockhash: Hash, fork: ForkId) {
+        let hash = self
+            .cache
+            .entry(blockhash)
+            .or_insert((fork, HashMap::new()));
+        hash.0 = std::cmp::max(fork, hash.0);
     }
 
-    /// copy the parents and parents' merges up to this instance, up to
-    ///   MAX_CACHE_ENTRIES deep
-    pub fn squash<U>(&mut self, parents: &[U])
-    where
-        U: Deref<Target = Self>,
-    {
-        for parent in parents {
-            if self.squash_parent_is_full(parent) {
-                break;
+    pub fn add_root(&mut self, fork: ForkId) {
+        self.roots.insert(fork);
+        if self.roots.len() > MAX_CACHE_ENTRIES {
+            if let Some(min) = self.roots.iter().min().cloned() {
+                self.roots.remove(&min);
+                self.cache.retain(|_, (fork, _)| *fork > min);
             }
         }
     }
 
-    /// Crate a new cache, pushing the old cache into the merged queue
-    pub fn new_cache(&mut self, blockhash: &Hash) {
-        assert!(self.active.is_some());
-        let merge = self.active.replace(Status::new(blockhash));
-
-        self.merges.push_front(Arc::new(merge.unwrap()));
-        if self.merges.len() > MAX_CACHE_ENTRIES {
-            self.merges.pop_back();
-        }
+    /// Insert a new signature for a specific fork.
+    pub fn insert(&mut self, transaction_blockhash: &Hash, sig: &Signature, fork: ForkId, res: T) {
+        let sig_map = self
+            .cache
+            .get_mut(transaction_blockhash)
+            .expect("new map should be created with register_hash");
```

Comment:
Could other validators potentially send/forward you a blockhash for a slot on a different fork that you haven't observed?

Reply:
nvm, I see that should be filtered out by the check_age() check earlier in the pipeline

Reply:
I'm seeing this assert on testnet. Is it possible that this insert check happens between these two lines: https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L433-L434, such that the check_age filter earlier in the pipeline passes, but then this assert is still triggered?

Reply:
That’s unexpected.  I can get rid of the assert.   The latest fork referencing the block hash can be the fork id for the map.

Reply:
Did you guys know that a panic while holding a read or or write guard will cause cascading panics to anyone else waiting on that lock?

Reply:
@garious yea, i know.  which is pretty cool way to track who is holding locks.